zio-sql/jvm/src/main/scala/Sql.scala

```diff
@@ -545,50 +569,103 @@ trait Sql {
-    sealed case class Literal[B: TypeTag](value: B)                   extends Expr[Features.Literal, Any, B]
+    def binary[F1, F2, A, B](left: Expr[F1, A, B], right: Expr[F2, A, B], op: BinaryOp[B]): Expr[F1 :||: F2, A, B] =
+      Expr(Expr.BinaryCase(left, right, op))
 
-    sealed case class AggregationCall[F, A, B, Z](param: Expr[F, A, B], aggregation: AggregationDef[B, Z])
-        extends Expr[Features.Aggregated[F], A, Z]
+    def functionCall1[F, A, B, Z](param: Expr[F, A, B], function: FunctionDef[B, Z]): Expr[F, A, Z] =
+      Expr(Expr.FunctionCall1Case(param, function))
 
-    sealed case class FunctionCall1[F, A, B, Z](param: Expr[F, A, B], function: FunctionDef[B, Z]) extends Expr[F, A, Z]
-
-    sealed case class FunctionCall2[F1, F2, A, B, C, Z](
+    def functionCall2[F1, F2, A, B, C, Z](
       param1: Expr[F1, A, B],
       param2: Expr[F2, A, C],
       function: FunctionDef[(B, C), Z]
-    ) extends Expr[Features.Union[F1, F2], A, Z]
+    ): Expr[F1 :||: F2, A, Z] =
+      Expr(Expr.FunctionCall2Case(param1, param2, function))
 
-    sealed case class FunctionCall3[F1, F2, F3, A, B, C, D, Z](
+    def functionCall3[F1, F2, F3, A, B, C, D, Z](
       param1: Expr[F1, A, B],
       param2: Expr[F2, A, C],
       param3: Expr[F3, A, D],
       function: FunctionDef[(B, C, D), Z]
-    ) extends Expr[Features.Union[F1, Features.Union[F2, F3]], A, Z]
+    ): Expr[F1 :||: F2 :||: F3, A, Z] =
+      Expr(Expr.FunctionCall3Case(param1, param2, param3, function))
 
-    sealed case class FunctionCall4[F1, F2, F3, F4, A, B, C, D, E, Z](
+    def functionCall4[F1, F2, F3, F4, A, B, C, D, E, Z](
       param1: Expr[F1, A, B],
       param2: Expr[F2, A, C],
       param3: Expr[F3, A, D],
       param4: Expr[F4, A, E],
       function: FunctionDef[(B, C, D, E), Z]
-    ) extends Expr[Features.Union[F1, Features.Union[F2, Features.Union[F3, F4]]], A, Z]
+    ): Expr[F1 :||: F2 :||: F3 :||: F4, A, Z] =
+      Expr(Expr.FunctionCall4Case(param1, param2, param3, param4, function))
+
+    def in[F, A, B](value: Expr[F, A, B], set: Read[B]): Expr[F, A, Boolean] =
+      Expr(Expr.InCase(value, set))
+
+    implicit def literal[A: TypeTag](a: A): Expr[Features.Literal, Any, A] =
+      Expr(Expr.LiteralCase(a))
+
+    def relational[F1, F2, A, B](
+      left: Expr[F1, A, B],
+      right: Expr[F2, A, B],
+      op: RelationalOp
+    ): Expr[F1 :||: F2, A, Boolean] =
+      Expr(Expr.RelationalCase(left, right, op))
+
+    private[sql] def source[A, B](tableName: TableName, column: Column[B]): Expr[Features.Source, A, B] =
+      Expr(Expr.SourceCase(tableName, column))
+
+    trait ExprCase[F, -A, +B, +T]
+
+    sealed case class AggregationCallCase[F, A, B, Z, T](param: T, aggregation: AggregationDef[B, Z])
+        extends ExprCase[Features.Aggregated[F], A, Z, T]
+
+    sealed case class BinaryCase[F1, F2, A, B, T](left: T, right: T, op: BinaryOp[B])
```

Comment:
I'm concerned that because of the need to use existentials in the recursive type (`Expr`), we have lost some type safety: we have no idea there is any connection between `T` and `B`.

(Similarly elsewhere, but this is a nice representative example.)

You can also see this happening not just with `B`, the output of the expression, but with features `F` too. 

I think to "solve" this problem we would have to make `T : [*, *] => *`, e.g. `T[F, B]` (assuming the source won't change in any recursive position — which may not be valid, need to think about that). This way here you could plug in `T[F1, B]` for left and `T[F2, B]` for right.

Then to regain fold, we could use `type Fold[F, B] = [Z] =>> Z`. 

Unfortunately the aesthetics / usability of this is rather terrible. Maybe this is a reason to not go down the recursion schemes path?

Reply:
What do you think of this? The internal implementation is fairly ugly but it seems like it creates a relatively nice API, at least given the complexity of the underlying data.